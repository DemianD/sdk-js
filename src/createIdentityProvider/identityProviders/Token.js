import { createIdentityProviderError } from '../../util/createError'

import {
  MISCONFIGURED,
  MISCONFIGURED_TOKEN_TYPE,
  MISCONFIGURED_ACCESS_TOKEN,
  MISCONFIGURED_REALM,
} from '../../constants'

export default class Token {
  constructor(network, { token_type, access_token, realm } = {}) {
    this._options = { token_type, access_token, realm }
    this._network = network

    if (!token_type) {
      throw createIdentityProviderError(
        'token_type is a required option for `Token`',
        MISCONFIGURED,
        MISCONFIGURED_TOKEN_TYPE
      )
    }

    if (!access_token) {
      throw createIdentityProviderError(
        'access_token is a required option for `Token`',
        MISCONFIGURED,
        MISCONFIGURED_ACCESS_TOKEN
      )
    }

    if (!realm) {
      throw createIdentityProviderError(
        'realm is a required option for `Token`',
        MISCONFIGURED,
        MISCONFIGURED_REALM
      )
    }
  }

  getAuthorization() {
    const { token_type, access_token, realm } = this._options

    if (this._current === undefined) {
      this._current = Promise.resolve({
        Authorization: `${token_type} ${access_token}`,
        Realm: realm,
      }).then(({ Realm, Authorization }) =>
        this._network
          .get(`${Realm}/integrations/proxy`, {
            headers: { Authorization },
          })
          .then(({ data: response }) => ({
            Realm: response.data.url,
            Authorization,
          }))
      )
    }

    return this._current
  }
}
