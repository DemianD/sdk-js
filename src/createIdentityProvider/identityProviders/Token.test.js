/* eslint-disable max-nested-callbacks */
import moxios from 'moxios'
import axios from 'axios'

import Token from './Token'

describe('Identity Provider # Token', () => {
  let network
  beforeEach(() => {
    network = axios.create()
    moxios.install(network)
  })

  afterEach(() => {
    moxios.uninstall(network)
  })

  it('should be able to create an identity provider instance', () => {
    const client = new Token(network, {
      token_type: 'Bearer',
      access_token: 'some-access-token-goes-here',
      realm: 'https://example.com',
    })

    expect(client).toBeDefined()
  })

  it('should error when no options are given', () => {
    expect(() => new Token(network)).toThrowErrorMatchingSnapshot()
  })

  it('should error when token_type is not given', () => {
    expect(
      () =>
        new Token(network, {
          access_token: 'some-access-token-goes-here',
          realm: 'https://example.com',
        })
    ).toThrowErrorMatchingSnapshot()
  })

  it('should error when access_token is not given', () => {
    expect(
      () =>
        new Token(network, {
          token_type: 'Bearer',
          realm: 'https://example.com',
        })
    ).toThrowErrorMatchingSnapshot()
  })

  it('should error when realm is not given', () => {
    expect(
      () =>
        new Token(network, {
          token_type: 'Bearer',
          access_token: 'some-access-token-goes-here',
        })
    ).toThrowErrorMatchingSnapshot()
  })

  it('should be possible to authenticate', async () => {
    const client = new Token(network, {
      token_type: 'Bearer',
      access_token: 'some-access-token-goes-here',
      realm: 'https://example.com',
    })

    // Mock the proxy call
    moxios.wait(() => {
      moxios.requests.mostRecent().respondWith({
        status: 200,
        response: {
          data: {
            url: 'https://api.example.com',
          },
        },
      })
    }, 0)

    // Wait for our call
    await client.getAuthorization()

    const {
      data,
      headers,
      method,
      params,
      url,
    } = moxios.requests.mostRecent().config

    expect({ data, headers, method, params, url }).toMatchSnapshot()

    // A call to auth + proxy should have been made
    expect(moxios.requests.count()).toBe(1)
  })

  it('should be possible to re-use the same auth request', async () => {
    const client = new Token(network, {
      token_type: 'Bearer',
      access_token: 'some-access-token-goes-here',
      realm: 'https://example.com',
    })

    // Mock the proxy call
    moxios.wait(() => {
      moxios.requests.mostRecent().respondWith({
        status: 200,
        response: {
          data: {
            url: 'https://api.example.com',
          },
        },
      })
    }, 0)

    // Wait for our call
    await client.getAuthorization()
    await client.getAuthorization()
    await client.getAuthorization()

    // Only the first auth + proxy call should have been made
    expect(moxios.requests.count()).toBe(1)
  })
})
