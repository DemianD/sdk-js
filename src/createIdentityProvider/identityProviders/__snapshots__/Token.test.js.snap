// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Identity Provider # Token should be possible to authenticate 1`] = `
Object {
  "data": undefined,
  "headers": Object {
    "Accept": "application/json, text/plain, */*",
    "Authorization": "Bearer some-access-token-goes-here",
  },
  "method": "get",
  "params": undefined,
  "url": "https://example.com/integrations/proxy",
}
`;

exports[`Identity Provider # Token should error when access_token is not given 1`] = `"[IDENTITY PROVIDER]: access_token is a required option for \`Token\`"`;

exports[`Identity Provider # Token should error when no options are given 1`] = `"[IDENTITY PROVIDER]: token_type is a required option for \`Token\`"`;

exports[`Identity Provider # Token should error when realm is not given 1`] = `"[IDENTITY PROVIDER]: realm is a required option for \`Token\`"`;

exports[`Identity Provider # Token should error when token_type is not given 1`] = `"[IDENTITY PROVIDER]: token_type is a required option for \`Token\`"`;
